README.md
DevSec PHP Baseline
===================

This is currently in development! Any feedback or Pull Request is welcome.

## Standalone Usage

This Compliance Profile requires [InSpec](https://github.com/chef/inspec) for execution:

```
$ git clone https://github.com/dev-sec/php-baseline
$ inspec exec php-baseline
```

You can also execute the profile directly from Github:

```
$ inspec exec https://github.com/dev-sec/php-baseline
```

## License and Author

* Author:: Dominik Richter <dominik.richter@googlemail.com>
* Author:: Christoph Hartmann <chris@lollyrock.com>

* Copyright 2015-2016, The Hardening Framework Team

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.


CHANGELOG.md
# Changelog

## [1.0.5](https://github.com/dev-sec/php-baseline/tree/1.0.5) (2023-05-02)

[Full Changelog](https://github.com/dev-sec/php-baseline/compare/1.0.4...1.0.5)

**Implemented enhancements:**

- use centralised issue templates and workflows [\#18](https://github.com/dev-sec/php-baseline/pull/18) ([schurzi](https://github.com/schurzi))

**Merged pull requests:**

- add spellchecking with codespell [\#20](https://github.com/dev-sec/php-baseline/pull/20) ([schurzi](https://github.com/schurzi))
- Configure Renovate [\#19](https://github.com/dev-sec/php-baseline/pull/19) ([renovate[bot]](https://github.com/apps/renovate))
- Change linting to Cookstyle [\#17](https://github.com/dev-sec/php-baseline/pull/17) ([schurzi](https://github.com/schurzi))

## [1.0.4](https://github.com/dev-sec/php-baseline/tree/1.0.4) (2022-01-12)

[Full Changelog](https://github.com/dev-sec/php-baseline/compare/1.0.3...1.0.4)

**Merged pull requests:**

- Minimum requirements [\#16](https://github.com/dev-sec/php-baseline/pull/16) ([micheelengronne](https://github.com/micheelengronne))
- fix rubocop error for Rakefile [\#14](https://github.com/dev-sec/php-baseline/pull/14) ([schurzi](https://github.com/schurzi))
- add dependency to chef-config for CI [\#13](https://github.com/dev-sec/php-baseline/pull/13) ([schurzi](https://github.com/schurzi))
- use version tag for changelog action [\#12](https://github.com/dev-sec/php-baseline/pull/12) ([schurzi](https://github.com/schurzi))
- Fix lint [\#11](https://github.com/dev-sec/php-baseline/pull/11) ([schurzi](https://github.com/schurzi))
- GitHub action [\#10](https://github.com/dev-sec/php-baseline/pull/10) ([rndmh3ro](https://github.com/rndmh3ro))

## [1.0.3](https://github.com/dev-sec/php-baseline/tree/1.0.3) (2020-07-23)

[Full Changelog](https://github.com/dev-sec/php-baseline/compare/1.0.2...1.0.3)

**Merged pull requests:**

- The release draft references the correct SHA [\#9](https://github.com/dev-sec/php-baseline/pull/9) ([micheelengronne](https://github.com/micheelengronne))

## [1.0.2](https://github.com/dev-sec/php-baseline/tree/1.0.2) (2020-06-18)

[Full Changelog](https://github.com/dev-sec/php-baseline/compare/1.0.1...1.0.2)

**Merged pull requests:**

- github actions release [\#8](https://github.com/dev-sec/php-baseline/pull/8) ([micheelengronne](https://github.com/micheelengronne))
- Update issue templates [\#7](https://github.com/dev-sec/php-baseline/pull/7) ([rndmh3ro](https://github.com/rndmh3ro))
- use recommended spdx license identifier [\#6](https://github.com/dev-sec/php-baseline/pull/6) ([chris-rock](https://github.com/chris-rock))
- spell correction in disable\_functions [\#5](https://github.com/dev-sec/php-baseline/pull/5) ([TLoebner](https://github.com/TLoebner))

## [1.0.1](https://github.com/dev-sec/php-baseline/tree/1.0.1) (2017-05-08)

[Full Changelog](https://github.com/dev-sec/php-baseline/compare/1.0.0...1.0.1)

**Merged pull requests:**

- update metadata [\#4](https://github.com/dev-sec/php-baseline/pull/4) ([chris-rock](https://github.com/chris-rock))
- restrict ruby testing to version 2.3.3 and update gemfile [\#3](https://github.com/dev-sec/php-baseline/pull/3) ([atomic111](https://github.com/atomic111))

## [1.0.0](https://github.com/dev-sec/php-baseline/tree/1.0.0) (2016-12-22)

[Full Changelog](https://github.com/dev-sec/php-baseline/compare/0.1.0...1.0.0)

**Merged pull requests:**

- migrate from Serverspec to InSpec [\#2](https://github.com/dev-sec/php-baseline/pull/2) ([chris-rock](https://github.com/chris-rock))

## [0.1.0](https://github.com/dev-sec/php-baseline/tree/0.1.0) (2016-11-09)

[Full Changelog](https://github.com/dev-sec/php-baseline/compare/f0de42996b3299ac0e9dc2821c693612f6134cbb...0.1.0)

**Merged pull requests:**

- fix typo in license header [\#1](https://github.com/dev-sec/php-baseline/pull/1) ([chris-rock](https://github.com/chris-rock))



\* *This Changelog was automatically generated by [github_changelog_generator](https://github.com/github-changelog-generator/github-changelog-generator)*


default/serverspec/phpconf_spec.rb
# encoding: utf-8
#
# Copyright 2015, Dominik Richter
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#

require 'spec_helper'

describe 'PHP config parameters' do
  # Some things we don't do:
  # * safe_mode
  #   reason: deprecated
  #   see: http://php.net/manual/de/features.safe-mode.php

  # Base configuration

  context php_config('open_basedir') do
    its(:value) { should eq '/srv/http/:/home/:/tmp/:/usr/share/pear/:/usr/share/webapps/' }
  end

  # Time / Memory Quota

  context php_config('max_execution_time') do
    its(:value) { should eq 30 }
  end

  context php_config('max_input_nesting_level') do
    its(:value) { should eq 64 }
  end

  context php_config('memory_limit') do
    its(:value) { should eq '128M' }
  end

  context php_config('post_max_size') do
    its(:value) { should eq '8M' }
  end

  # PHP Capabilities

  # TODO: do we have a recommended minimum-set?
  context php_config('disable_functions') do
    its(:value) { should eq 'php_uname, getmyuid, getmypid, passthru, leak, listen, diskfreespace, tmpfile, link, ignore_user_abord, shell_exec, dl, set_time_limit, exec, system, highlight_file, source, show_source, fpaththru, virtual, posix_ctermid, posix_getcwd, posix_getegid, posix_geteuid, posix_getgid, posix_getgrgid, posix_getgrnam, posix_getgroups, posix_getlogin, posix_getpgid, posix_getpgrp, posix_getpid, posix, _getppid, posix_getpwnam, posix_getpwuid, posix_getrlimit, posix_getsid, posix_getuid, posix_isatty, posix_kill, posix_mkfifo, posix_setegid, posix_seteuid, posix_setgid, posix_setpgid, posix_setsid, posix_setuid, posix_times, posix_ttyname, posix_uname, proc_open, proc_close, proc_get_status, proc_nice, proc_terminate, phpinfo' }
  end

  # TODO: do we have a recommended minimum-set?
  context php_config('disable_classes') do
    its(:value) { should eq '...' }
  end

  context php_config('register_globals') do
    its(:value) { should eq 'Off' }
  end

  context php_config('expose_php') do
    its(:value) { should eq 'Off' }
  end

  context php_config('enable_dl') do
    its(:value) { should eq 'Off' }
  end

  context php_config('default_charset') do
    its(:value) { should eq 'utf-8' }
  end

  context php_config('default_mimetype') do
    its(:value) { should eq 'text/html' }
  end

  # removed as of PHP5.4, so...
  # remove these test?
  context php_config('magic_quotes_gpc') do
    its(:value) { should eq nil}
  end
  context php_config('magic_quotes_sybase') do
    its(:value) { should eq nil}
  end

  # # removed // how to test this?
  # context php_config('magic_quotes_runtime') do
  #   its(:value) { should eq 'Off'}
  # end

  # Upload / Open

  context php_config('allow_url_fopen') do
    its(:value) { should eq 'Off' }
  end

  context php_config('allow_url_include') do
    its(:value) { should eq 'Off' }
  end

  context php_config('file_uploads') do
    its(:value) { should eq 'Off' }
  end

  # Alternative: restrict upload maximum to prevent
  # system overload:
  # upload_tmp_dir = /var/php_tmp
  # upload_max_filezize = 10M

  # Log // Information Disclosure

  context php_config('display_errors') do
    its(:value) { should eq 'Off' }
  end

  context php_config('display_startup_errors') do
    its(:value) { should eq 'Off' }
  end

  context php_config('log_errors') do
    its(:value) { should eq 'On' }
  end

  # Session Handling

  context php_config('session.save_path') do
    its(:value) { should eq '/var/lib/php' }
  end

  context php_config('session.cookie_httponly') do
    its(:value) { should eq 1 }
  end

  # Mail

  context php_config('mail.add_x_header') do
    its(:value) { should eq 'Off' }
  end

end


libraries/php.rb
# frozen_string_literal: true

#
# Copyright:: 2016, Christoph Hartmann
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#

class PHP < Inspec.resource(1)
  name 'php'

  example "
    describe php.config('max_execution_time') do
      it { should cmp 0 }
    end

    describe php.extension('tillext') do
    	it { should_not be_loaded }
    end

    describe php.extension('session') do
      it { should be_loaded }
    end
  "

  def initialize(options = {})
    super()
    @options = options
  end

  # returns php runtime information
  def info
    inspec.command("#{_php_executable} -i")
  end

  def config(param)
    cmd = inspec.command("#{_php_executable} -r 'echo get_cfg_var( \"#{param}\" );'")
    cmd.stdout
  end

  def extension(extension)
    ext = Class.new do
      def initialize(parent, extension)
        super
        @parent = parent
        @extension = extension
      end

      def loaded?
        cmd = @parent._run("#{@parent._php_executable} --ri '#{@extension}'")
        return true if cmd.exit_status.zero?

        false
      end

      def to_s
        "PHP Extension #{@extension}"
      end
    end
    ext.new(self, extension)
  end

  # internal methods

  def _php_executable
    cmd = 'php'
    cmd += " -c #{@options[:ini]}" if !@options.nil? && @options.key?(:ini)
    cmd
  end

  def _run(cmd)
    inspec.command(cmd)
  end
end

# Deprecated Serverspec resource
class PHPConfig < PHP
  name 'php_config'

  example "
    describe php_config('default_mimetype') do
      its('value') { should eq 'text/html' }
    end
  "

  def initialize(param)
    super()
    warn '[DEPRECATION] `php_config(param)` is deprecated.  Please use `php.config(param)` instead.'
    @param = param
  end

  def value
    config(@param)
  end

  def to_s
    "PHP Config #{@param}"
  end
end

# Deprecated Serverspec resource
class PHPExtension < PHP
  name 'php_extension'

  example "
    describe php_extension('tillext') do
    	  it { should_not be_loaded }
    end

    describe php_extension('session') do
      it { should be_loaded }
    end
  "

  def initialize(extension)
    super
    warn '[DEPRECATION] `php_extension(ext)` is deprecated.  Please use `php.extension(ext)` instead.'
    @extension = extension
  end

  def loaded?
    extension(@extension).loaded?
  end

  def to_s
    "PHP Extension #{@extension}"
  end
end


default/serverspec/spec_helper.rb
# encoding: utf-8
#
# Copyright 2015, Dominik Richter
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#

if ENV['STANDALONE_SPEC']

  require 'serverspec'
  require 'pathname'
  require 'net/ssh'
  require 'highline/import'

  set :backend, :ssh

  RSpec.configure do |c|

    if ENV['ASK_SUDO_PASSWORD']
      c.sudo_password = ask('Enter sudo password: ') { |q| q.echo = false }
    else
      c.sudo_password = ENV['SUDO_PASSWORD']
    end

    options = {}

    if ENV['ASK_LOGIN_PASSWORD']
      options[:password] = ask("\nEnter login password: ") { |q| q.echo = false }
    else
      options[:password] = ENV['LOGIN_PASSWORD']
    end

    if ENV['ASK_LOGIN_USERNAME']
      options[:user] = ask("\nEnter login username: ") { |q| q.echo = false }
    else
      options[:user] = ENV['LOGIN_USERNAME'] || ENV['user'] || Etc.getlogin
    end

    if options[:user].nil?
      puts 'specify login user env LOGIN_USERNAME= or user='
      exit 1
    end

    c.host = ENV['TARGET_HOST']
    c.ssh_options = options.merge(Net::SSH::Config.for(c.host))

  end

else
  require 'serverspec'

  set :backend, :exec
end


controls/php_conf.rb
# frozen_string_literal: true

#
# Copyright:: 2015, Dominik Richter
# Copyright:: 2016, Christoph Hartmann
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#

title 'PHP config parameters'

# Some things we don't do:
# * safe_mode
#   reason: deprecated
#   see: http://php.net/manual/de/features.safe-mode.php

# Base configuration
describe php_config('open_basedir') do
  its('value') { should eq '/srv/http/:/home/:/tmp/:/usr/share/pear/:/usr/share/webapps/' }
end

# Time / Memory Quota

describe php_config('max_execution_time') do
  its('value') { should cmp 30 }
end

describe php_config('max_input_nesting_level') do
  its('value') { should cmp 64 }
end

describe php_config('memory_limit') do
  its('value') { should eq '128M' }
end

describe php_config('post_max_size') do
  its('value') { should eq '8M' }
end

# PHP Capabilities

# TODO: do we have a recommended minimum-set?
describe php_config('disable_functions') do
  its('value') { should eq 'php_uname, getmyuid, getmypid, passthru, leak, listen, diskfreespace, tmpfile, link, ignore_user_abort, shell_exec, dl, set_time_limit, exec, system, highlight_file, source, show_source, fpaththru, virtual, posix_ctermid, posix_getcwd, posix_getegid, posix_geteuid, posix_getgid, posix_getgrgid, posix_getgrnam, posix_getgroups, posix_getlogin, posix_getpgid, posix_getpgrp, posix_getpid, posix, _getppid, posix_getpwnam, posix_getpwuid, posix_getrlimit, posix_getsid, posix_getuid, posix_isatty, posix_kill, posix_mkfifo, posix_setegid, posix_seteuid, posix_setgid, posix_setpgid, posix_setsid, posix_setuid, posix_times, posix_ttyname, posix_uname, proc_open, proc_close, proc_get_status, proc_nice, proc_terminate, phpinfo' }
end

# TODO: do we have a recommended minimum-set?
describe php_config('disable_classes') do
  its('value') { should cmp '...' }
end

describe php_config('register_globals') do
  its('value') { should cmp 'Off' }
end

describe php_config('expose_php') do
  its('value') { should cmp 'Off' }
end

describe php_config('enable_dl') do
  its('value') { should cmp 'Off' }
end

describe php_config('default_charset') do
  its('value') { should cmp 'utf-8' }
end

describe php_config('default_mimetype') do
  its('value') { should cmp 'text/html' }
end

# removed as of PHP5.4, so...
# remove these test?
describe php_config('magic_quotes_gpc') do
  its('value') { should eq nil }
end
describe php_config('magic_quotes_sybase') do
  its('value') { should eq nil }
end

# # removed // how to test this?
# describe php_config('magic_quotes_runtime') do
#   its('value') { should eq 'Off'}
# end

# Upload / Open

describe php_config('allow_url_fopen') do
  its('value') { should cmp 'Off' }
end

describe php_config('allow_url_include') do
  its('value') { should cmp 'Off' }
end

describe php_config('file_uploads') do
  its('value') { should cmp 'Off' }
end

# Alternative: restrict upload maximum to prevent
# system overload:
# upload_tmp_dir = /var/php_tmp
# upload_max_filezize = 10M

# Log // Information Disclosure

describe php_config('display_errors') do
  its('value') { should cmp 'Off' }
end

describe php_config('display_startup_errors') do
  its('value') { should cmp 'Off' }
end

describe php_config('log_errors') do
  its('value') { should cmp 'On' }
end

# Session Handling
describe php_config('session.save_path') do
  its('value') { should cmp '/var/lib/php' }
end

describe php_config('session.cookie_httponly') do
  its('value') { should cmp 1 }
end

# Mail

describe php_config('mail.add_x_header') do
  its('value') { should cmp 'Off' }
end
